feat: add pub modifier for functions accessibility

guipublic opened the pull request:
# Description

## Problem\*

When implementing a library, a developper cannot expose public functions to be called externally and hide internal functions.

Resolves #776

## Summary\*

We add a'pub' modifier for functions and if a function calls a non 'pub' function from a different module, we report an error.
This is done in a new pass which is run after the type checking and before the monomorphisation.
We first collect the functions called from main and check their module and 'pub' modifier.
Then we recursively check these called functions, if not already processed.
In case of a contract, we check all the functions from the crate.

## Documentation

- [X] This PR requires documentation updates when merged.

  <!-- If checked, check one of the following: -->

  - [ ] I will submit a noir-lang/docs PR.

  <!-- Submit a PR on https://github.com/noir-lang/docs. Thank you! -->

  - [X] I will request for and support Dev Rel's help in documenting this PR.

  <!-- List / highlight what should be documented. -->
  <!-- Dev Rel will reach out for clarifications when needed. Thank you! -->
When designing a library, functions that can be called exgternally should have a 'pub' modifier.

## Additional Context

I have added 'pub' to the stdlib, but not in an exhaustive way.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

TomAFrench commented on crates/noirc_frontend/src/hir/mod.rs:
```diff
@@ -3,6 +3,7 @@ pub mod def_map;
 pub mod resolution;
 pub mod scope;
 pub mod type_check;
+pub mod visilibity;
```
```suggestion
pub mod visibility;
```

This module name has a typo

guipublic replied:
Done

jfecher reviewed (changes requested):
We shouldn't need another pass for this, it should be part of name resolution

TomAFrench commented on noir_stdlib/src/string.nr:
```diff
@@ -5,7 +5,7 @@ impl<N> str<N> {
     fn as_bytes(_self: Self) -> [u8; N] { }
```
Why is this not marked public? Similarly with slice methods.

guipublic replied:
I did the minimum changes so that the tests are passing. However a bunch of methods should need the 'pub' modifier. I did investigate and in fact I was no parsing the statements when looking for function calls.
Most of the stdlib is now 'pub'

TomAFrench commented on crates/noirc_frontend/src/hir/visilibity.rs:
```diff
@@ -0,0 +1,122 @@
+    }
+
+    /// Checks that the input function is only calling functions with 'pub' modifier outside its module
+    /// recusively checks the calling functions, if they are not already checked.
```
```suggestion
    /// recursively checks the calling functions, if they are not already checked.
```
nit

guipublic replied:
Done

TomAFrench reviewed (commented):
Nice, I've been looking forward to being able to have proper private functions. :tada:

Just a couple of small things

guipublic commented:
> We shouldn't need another pass for this, it should be part of name resolution

I put it after name resolution because I need all the names to be resolved when we reach a function call, in order to be able to check it properly.

guipublic commented:
A function called from a different module requires a 'pub' modifier, this means that even a sub-module in the same file requires the pub to be used in main. This is probably too restrictive, but since I am not sure of the exact behaviour that we want, I leave it like this for now.

jfecher commented:
> I put it after name resolution because I need all the names to be resolved when we reach a function call, in order to be able to check it properly.

The function name should be resolved when we reach the identifier expression referencing it in the name resolver. You'll have access to the function's definition in the node resolver and should be able to use that to check whether it is public.

jfecher commented on crates/noirc_frontend/src/hir/def_collector/dc_crate.rs:
```diff
@@ -228,6 +229,18 @@ impl DefCollector {
             def_collector.collected_traits_impls,
             errors,
         );
+        // check that a function is calling a pub function if from another module.
+        let mut visibility = FunctionVisibility::default();
+        for (_, func_id) in &file_func_ids {
+            visibility.check_visibility(context, func_id);
+        }
+        for (_, func_id) in &file_method_ids {
+            visibility.check_visibility(context, func_id);
+        }
+        for (_, func_id) in &file_trait_impls_ids {
+            visibility.check_visibility(context, func_id);
+        }
+        errors.extend(visibility.errors);
```
We don't need a separate pass to go through each function, collect all function ids that are called, and then check the visibility on those. We can do it within the name resolution pass when these variables are referred to here: https://github.com/noir-lang/noir/blob/master/crates/noirc_frontend/src/hir/resolution/resolver.rs#L1035

guipublic replied:
I can't retrieve the function metadata from resolver.rs#L1035: `all function ids should have metadata`, which makes sense to me because we are resolving functions and we do not guarantee it will be done in the correct order.

jfecher replied:
Can we add metadata or just whether a function is `pub` or not during def collection instead then? It'd make sense if we populated each FuncMeta during def collection then just had to check them during resolution.

jfecher replied:
I'm working on a branch now to move the pub check to name resolution instead. Since separating out which fields of FuncMeta we can resolve early from those we cant would be a larger change, I've just added a separate hash map now for mapping functions to their visibilities which is filled out early.

guipublic commented:
I moved the check to the name resolution but I still need to have the functions resolved.
I changed the error to warning, warning for deprecated in future.
I allow private calls from a submodule.

kevaundray commented:
superceded by #2754
